### ISSUE TITLE
Clone boot partitions via rsync, and use efibootmgr

### ISSUE THREAD

**Issue by saveriomiroddi** [saveriomiroddi/zfs-installer#7]
Apply feedback from G.Beekmans:

> Where your "dd" the two partitions works well, I find rsync is simply faster. This also keeps the two partition UUIDs unique, which may or may not be desirable. Right now it'll mount one of the two EFIs at random if using UUID mounting. It's not going to always be /dev/sda1. Often /dev/sdb1 mounted instead, which means a future 'dd' command will go the wrong direction if you don't test for this and catch it in code.
>
> Here's how I ended up making it work for me in a more deterministic way:
>
>     echo PARTUUID=$(blkid -s PARTUUID -o value $disk1_id-$efi) /boot/efi vfat defaults 0 0 >> /etc/fstab
>     echo PARTUUID=$(blkid -s PARTUUID -o value $disk2_id-$efi) /boot/efi2 vfat defaults 0 0 >> /etc/fstab
>
>     mkdir /boot/efi
>     mkdir /boot/efi2
>
>     mount /boot/efi
>     mount /boot/efi2
>
>     grub-install --target=x86_64-efi --efi-directory=/boot/efi --bootloader-id=ubuntu --recheck --no-floppy
>
>     rsync -Rai --stats --human-readable --delete --verbose --progress /boot/efi/./ /boot/efi2
>
>     efibootmgr --create --gpt --disk $disk2 --part 1 --label "ubuntu-disk2" --loader '\EFI\ubuntu\shimx64.efi'
>
> The "grub-install" command can actually be replaced with a efibootmgr command instead to make the two the same. Both end up doing the same thing so functionally they appear interchangeable. I think the efibootmgr is more flexible and is more easily understood vs hiding behind "grub-install".
>
> In this instance $disk2 is a variable I setup to refer to the appropriate /dev/disk/by-id/ link.

With this solutions, all the EFI partitions are mounted as `/boot/efi<N>` (the first without <N>). If the first device breaks, boot-related updates won't find `/boot/efi` and may break, but that's tolerable/unavoidable - in this case, the sysadmin should replace the device and restore the partition/fstab.

Note that `nofail,x-systemd.device-timeout=1` mount flags should be used (see [Arch wiki](https://wiki.archlinux.org/index.php/fstab#External_devices)); above is only `defaults`, which should cause the boot to halt if a partition is not present (does this doesn't happen to Beekmans/kongkrit?).

**Comment by kongkrit** [comment]
@saveriomiroddi ,

Hi Again,
I would like to check whether I understand this correctly:
- If we used `dd` to copy disk0-part1 to disk1-part1, both disk0-part1 and disk1-part1 will have the same UUID.
- Therefore, when we do an `update-grub`, only one partition will be written to, and you won't know which one. Therefore, disk0-part1 and disk1-part1 will be out of sync.
- And we also cannot guarantee which one will be used to do EFI boot? so `update-grub` may not take effect if it chose the un-updated partition?

- If we make this modification, then we want to do update-grub, do we have to specify a disk partition, or will dpkg triiger rsync will magically keep the partition that has more up-to-date files as the source and propagate that to the other partition?

**Comment by saveriomiroddi** [comment]
> @saveriomiroddi,
> 
> I have made the above changes myself (for ubuntu only) at https://github.com/kongkrit/zfs-installer -- but i haven't tested it. if it was successful, i will let you know and i may ask for your help if this could be a pull request into this repo.

OK, then I'll wait any notification (or PR) from you :smile:. In the latter case, don't forget to check the build. Based on a quick check, there's at least one change in order to make shellcheck happy :smile:

**Comment by kongkrit** [comment]
> OK, then I'll wait any notification (or PR) from you 😄. In the latter case, don't forget to check the build. Based on a quick check, there's at least one change in order to make shellcheck happy 😄

@saveriomiroddi,

I have tested the script on Ubuntu18.04.3 desktop and it works under VirtualBox. After installation, I unmounted both disks (one at a time, of course), reboot and the system came up with degraded zpools.

I haven't had a chance to test it on debian yet, and i haven't made the changes in the debian related functions, but the code in my forked repo is now shellcheck 0.7.0 clean.

Now please educate me a bit. Without swap, we have 3 partitions, EFI, bpool, rpool. bpool and rpool are ZFS. My understanding is that EFI partitions will only get written when you do a grub update. Is that correct? So we need to have `update-grub` trigger an rsync to other EFI partitions, making sure every EFI partition in the mirror set have the latest files. Am I having the right idea, now that each EFI partition has a distinct UUID?

**Comment by saveriomiroddi** [comment]
Hello! It'll take me a bit of time to participate to this thread, because I have a presentation at the preFOSDEM next week, so I'm fully focused on it. I'll notify you a few days after I'm back.

**Comment by kongkrit** [comment]
@saveriomiroddi,

Definitely take your time. I wish you the best on your presentation. Take care and we can converse afterwards. Thanks for all the help so far.

**Comment by kongkrit** [comment]
@saveriomiroddi,

I got the script shellcheck 0.7.0 clean, and it's working with ubuntu desktop 18.04.3 live and ubuntu 18.04.3 server. I cannot, however, get it to work with Debian 10.2. I noticed you used Debian 10.1 for testing. I tried many Debian 10.2 images, but none launched the installer successfully.

**Comment by saveriomiroddi** [comment]
> @saveriomiroddi,
> 
> I got the script shellcheck 0.7.0 clean, and it's working with ubuntu desktop 18.04.3 live and ubuntu 18.04.3 server. I cannot, however, get it to work with Debian 10.2. I noticed you used Debian 10.1 for testing. I tried many Debian 10.2 images, but none launched the installer successfully.

Hello! I've resume the work on my open source projects. I'm handling the Debian problem first (see [new issue](https://github.com/saveriomiroddi/zfs-installer/issues/55), then I'll get back to this PR.

**Comment by saveriomiroddi** [comment]
@kongkrit can you open a PR? I'll start the review process.

**Comment by saveriomiroddi** [comment]
> Hi Again,
> I would like to check whether I understand this correctly:

I didn't answer your questions before, apologies 😬

So!

> If we used dd to copy disk0-part1 to disk1-part1, both disk0-part1 and disk1-part1 will have the same UUID.
> Therefore, when we do an update-grub, only one partition will be written to, and you won't know which one. Therefore, disk0-part1 and disk1-part1 will be out of sync.

Correct.

> And we also cannot guarantee which one will be used to do EFI boot? so update-grub may not take effect if it chose the un-updated partition?

Based on my knowledge:

- via efibootmgr, the EFI partition are registered in the EFI firmware (say, the "BIOS")
- the firmware has a preset ordering, that the user can change
- very likely, without user changing anything, the order of registration is applied
- therefore, the first is loaded

> If we make this modification, then we want to do update-grub, do we have to specify a disk partition, or will dpkg triiger rsync will magically keep the partition that has more up-to-date files as the source and propagate that to the other partition?
> Will kernel update also has this problem?

So, with `dd`, changes are going to a random EFI partition, since, due to the EFI partition UUIDS being the same, the mounted partition will be chosen non-deterministically.

This is, however, a related-but-different topic from using `rsync` vs `dd`. An explanation of mine is [here](https://saveriomiroddi.github.io/Installing-Ubuntu-on-a-ZFS-root-with-encryption-and-mirroring/#comment-4750203350).

The positive side is that there should be a unified approach to this - a dpkg hook. Additionally, changes that would break an installation are unlikely, possibly, not real-world cases - the kernel images still reside on `/boot`, which is a ZFS partition, which is always in sync because it's mirrored.

**Comment by saveriomiroddi** [comment]
Other answers 😄

> Without swap, we have 3 partitions, EFI, bpool, rpool. bpool and rpool are ZFS. My understanding is that EFI partitions will only get written when you do a grub update. Is that correct? So we need to have update-grub trigger an rsync to other EFI partitions, making sure every EFI partition in the mirror set have the latest files. Am I having the right idea, now that each EFI partition has a distinct UUID?

If I understand correctly, yes and no 😄

Let's suppose we update the kernel. This will add kernel images to `/boot` and will update the GRUB configuration.

Now, let's check on a real-world configuration:

```
root@c64:/boot# ls -l
-rw-r--r-- 1 root root   235811 Dec 18 16:01 config-5.3.0-26-generic
-rw-r--r-- 1 root root   235811 Jan 15 19:10 config-5.3.0-28-generic
drwxr-xr-x 5 root root     4096 Jan  1  1970 efi
drwxr-xr-x 6 root root     4096 Feb  3 17:40 grub
-rw-r--r-- 1 root root 68406172 Jan 24 08:54 initrd.img-5.3.0-26-generic
-rw-r--r-- 1 root root 68406980 Feb  3 17:40 initrd.img-5.3.0-28-generic
-rw-r--r-- 1 root root   182704 Jan 28  2016 memtest86+.bin
-rw-r--r-- 1 root root   184380 Jan 28  2016 memtest86+.elf
-rw-r--r-- 1 root root   184840 Jan 28  2016 memtest86+_multiboot.bin
-rw------- 1 root root  4487587 Dec 18 16:01 System.map-5.3.0-26-generic
-rw------- 1 root root  4487587 Jan 15 19:10 System.map-5.3.0-28-generic
-rw------- 1 root root  9146616 Dec 18 16:20 vmlinuz-5.3.0-26-generic
-rw------- 1 root root  9146616 Jan 15 22:28 vmlinuz-5.3.0-28-generic

root@c64:/boot# ag 5.3.0-28
grub/grub.cfg
151:        linux /boot/vmlinuz-5.3.0-28-generic root=UUID=25440acc-3120-4810-8792-33c6ed9d5ab9 ro  amd_iommu=on ipv6.disable=1
152:  initrd  /boot/initrd.img-5.3.0-28-generic
155:  menuentry 'Linux Mint 19.3 MATE, with Linux 5.3.0-28-generic' --class linuxmint --class gnu-linux --class gnu --class os $menuentry_id_option 'gnulinux-5.3.0-28-generic-advanced-25440acc-3120-4810-8792-33c6ed9d5ab9' {
169:    echo  'Loading Linux 5.3.0-28-generic ...'
170:          linux /boot/vmlinuz-5.3.0-28-generic root=UUID=25440acc-3120-4810-8792-33c6ed9d5ab9 ro  amd_iommu=on ipv6.disable=1
172:    initrd  /boot/initrd.img-5.3.0-28-generic
174:  menuentry 'Linux Mint 19.3 MATE, with Linux 5.3.0-28-generic (recovery mode)' --class linuxmint --class gnu-linux --class gnu --class os $menuentry_id_option 'gnulinux-5.3.0-28-generic-recovery-25440acc-3120-4810-8792-33c6ed9d5ab9' {
187:    echo  'Loading Linux 5.3.0-28-generic ...'
188:          linux /boot/vmlinuz-5.3.0-28-generic root=UUID=25440acc-3120-4810-8792-33c6ed9d5ab9 ro recovery nomodeset
190:    initrd  /boot/initrd.img-5.3.0-28-generic

config-5.3.0-28-generic
3:# Linux/x86 5.3.0-28-generic Kernel Configuration
41:CONFIG_VERSION_SIGNATURE="Ubuntu 5.3.0-28.30~18.04.1-generic 5.3.13"
```

as you see, the kernel images related information is *not* in not in the EFI partition, but in the boot partition.

So when a kernel is updated, even if the EFI partitions are not synced, the (default) kernel loaded will still be the latest one.

Now, why would we want to keep the EFI partitions in sync? At least two cases I can think of:

1. another operating system is added, say, Windows: this will add an entry to `/boot/efi`
2. GRUB itself is updated: again, this modifies `/boot/efi`

Additionally, assuming that the partition written is always the same (via `rsync`), in order to cause a problem, the *first* device needs to break. If the second breaks, but the first is always the loaded one, the updates will be loaded.

Based on the above, the problems are two:

1. the EFI partition mounted as /boot/efi is non-deterministic (solved via `rsync`)
2. the EFI partitions are not kept in sync

your coming PR will address problem 1.; problem 2. will be addresses separately. As stated above, the two cases are not blockers, so it's not something that needs to be addressed immediately.

**Comment by saveriomiroddi** [comment]
@kongkrit some other interesting notes: https://github.com/saveriomiroddi/zfs-installer/issues/60.

**Comment by kongkrit** [comment]
HI @saveriomiroddi,

Sorry for late reply. I was busy with work. I read all of the above, and I think I got most (but probably not all) of it. If I understand correctly, we still have a small chance that the system might "break" and that is:

- The designated boot drive (by EFI BIOS) is always the same, and if that one breaks, system won't boot. We have to go into BIOS and force it to boot from another one of our mirrors.

- We still don't have an automated way to rsync the EFI partition every time that boot drive EFI partition is modified.

**Comment by kongkrit** [comment]
@saveriomiroddi ,

Given all the above, it seems like I should hold my PR on "rsync EFI partitions, so each parition has unique UUID". correct?

**Comment by saveriomiroddi** [comment]
> The designated boot drive (by EFI BIOS) is always the same, and if that one breaks, system won't boot. We have to go into BIOS and force it to boot from another one of our mirrors.

I don't think this is the way it happens, rather, the BIOS has a list of devices from whom to boot.

If the BIOS had a fixed boot device, it wouldn't boot, say, from a USB key, if it was inserted.

**Comment by saveriomiroddi** [comment]
> @saveriomiroddi ,
> 
> Given all the above, it seems like I should hold my PR on "rsync EFI partitions, so each parition has unique UUID". correct?

Yes, correct.

**Comment by saveriomiroddi** [comment]
> * We still don't have an automated way to rsync the EFI partition every time that boot drive EFI partition is modified.

Yes, this is a [separate issue](#60).

**Comment by saveriomiroddi** [comment]
> HI @saveriomiroddi,
> 
> Sorry for late reply. I was busy with work. I read all of the above, and I think I got most (but probably not all) of it. If I understand correctly, we still have a small chance that the system might "break" and that is:
> 
> * The designated boot drive (by EFI BIOS) is always the same, and if that one breaks, system won't boot. We have to go into BIOS and force it to boot from another one of our mirrors.
> * We still don't have an automated way to rsync the EFI partition every time that boot drive EFI partition is modified.

@kongkrit I haven't heard from you from a long time :smile: Since the feature was easy, and it's been requested, I've followed up and implemented it (#68). If you'd still like to contribute, #60 is a similar, related addition, but there are also other interesting & easy issues.

**Comment by kongkrit** [comment]
Hi @saveriomiroddi,

I was holding it back since it seemed like I did not get a clear conclusion (most possibly my misunderstanding) whether we should have different UUIDs for different drive's EFI's partition or not. Seems like you went ahead and implement different UUIDs for different drives now.

My "linux-related-knowledge" is still a very noob level (it's so embarassing, since I first used it in 1995-96 or so with LILO and bought the early Slackware/RedHat CDs! -- got excited waiting for 1.2.0 kernel release haha!)

Right now I am fairly busy and cannot have a clear time frame of when I can come around and learn enough to fix [#60](https://github.com/saveriomiroddi/zfs-installer/issues/60). If an expert like you could work on it, please go ahead.

Just want to express my appreciation again with your work. **Thank you.** It got me in love with ZFS and I'm now running 2 experimental VMs with focal ubuntu-minimal dev branch. I have created some (much less beautiful than yours) [scripts](https://github.com/kongkrit/zfs-ouds) to install via debootstrap to get a minimal system.

**Comment by saveriomiroddi** [comment]
> Seems like you went ahead and implement different UUIDs for different drives now.

Yes, at some point I've realized this is actually _very_ important. Having the same UUIDs causes changes to go to a random drive (due to non-deterministic mount), which is highly undesirable. I don't think I will add back extra support for this kind of setup, since I can't find a use case for it.

Thanks for your kind words. You can always have a look to the issues in the future. I'm very welcoming towards PRs, so even if you think something is broken, open a PR, you never know :smile: The only strict aspect is that any given PR should accomplish only one task (so that it's easy to work on it) :slightly_smiling_face:

**Comment by kongkrit** [comment]
> > Seems like you went ahead and implement different UUIDs for different drives now.
> 
> Yes, at some point I've realized this is actually _very_ important. Having the same UUIDs causes changes to go to a random drive (due to non-deterministic mount), which is highly undesirable. I don't think I will add back extra support for this kind of setup, since I can't find a use case for it.

@saveriomiroddi ,
Thanks for the update. The more I think about it, the more I think ESPs on drives should have their own UUIDs.

The update, though, should be fairly minimal, isn't it? I am not too knowledgable about grub, but since grub can mount zfs /boot, are not most of the changes there in /boot and not /boot/efi?

```
:/boot/efi/EFI/ubuntu$ ls -l
total 3644
-rwxr-xr-x 1 root root     108 Jan  3 23:52 BOOTX64.CSV
-rwxr-xr-x 1 root root     112 Jan  3 23:52 grub.cfg
-rwxr-xr-x 1 root root 1116024 Jan  3 23:52 grubx64.efi
-rwxr-xr-x 1 root root 1269496 Jan  3 23:52 mmx64.efi
-rwxr-xr-x 1 root root 1334816 Jan  3 23:52 shimx64.efi
```
only `grub.cfg` and `BOOTX64.csv` are there. grub.cfg is extremely simple:

```
/boot/efi/EFI/ubuntu$ more grub.cfg
search.fs_uuid 1ee6d9d5101de731 root hd0,gpt2 hd1,gpt2
set prefix=($root)'/@/grub'
configfile $prefix/grub.cfg
```

**Comment by saveriomiroddi** [comment]
> The more I think about it, the more I think ESPs on drives should have their own UUIDs.
> 
> The update, though, should be fairly minimal, isn't it?

If I understand correctly your question, the answer is: they do now, after #68 was merged :smile:.

The way it was working before it was:

1. the partitions are [created](https://github.com/saveriomiroddi/zfs-installer/blob/25dc6781b86a3359ab51776720f9835312f1aa8d/install-zfs.sh#L579) and [formatted](https://github.com/saveriomiroddi/zfs-installer/blob/25dc6781b86a3359ab51776720f9835312f1aa8d/install-zfs.sh#L612-L614)
1. then they are [synced via dd](https://github.com/saveriomiroddi/zfs-installer/blob/bb9c1f018ab58d990467b8c98064b391c3d04d4f/install-zfs.sh#L1057-L1060)

After 1., they have different UUIDS, however 2. will make them all have the same.

With the change by #68, which, on step 2., performs [rsync instead of dd](https://github.com/saveriomiroddi/zfs-installer/pull/68/files#diff-761944bb90b658235bf6861db38b75e7R1061), so that the UUIDs are not changed to be the same :slightly_smiling_face:.
